ext/semian/sysv_semaphores.c

```diff
@@ -136,8 +132,92 @@ acquire_semaphore(void *p)
 {
   semian_resource_t *res = (semian_resource_t *) p;
   res->error = 0;
+#ifdef DEBUG
+  print_sem_vals(res->sem_id);
+#endif
   if (perform_semop(res->sem_id, SI_SEM_TICKETS, -1, SEM_UNDO, &res->timeout) == -1) {
     res->error = errno;
   }
   return NULL;
 }
+
+static key_t
+generate_key(const char *name)
+{
+  char semset_size_key[20];
+  char *uniq_id_str;
+
+  // It is necessary for the cardinatily of the semaphore set to be part of the key
+  // or else sem_get will complain that we have requested an incorrect number of sems
+  // for the desired key, and have changed the number of semaphores for a given key
+  sprintf(semset_size_key, "_NUM_SEMS_%d", SI_NUM_SEMAPHORES);
+  uniq_id_str = malloc(strlen(name)+strlen(semset_size_key)+1);
+  strcpy(uniq_id_str, name);
+  strcat(uniq_id_str, semset_size_key);
+
+  union {
+    unsigned char str[SHA_DIGEST_LENGTH];
+    key_t key;
+  } digest;
+  SHA1((const unsigned char *) uniq_id_str, strlen(uniq_id_str), digest.str);
+  free(uniq_id_str);
+  /* TODO: compile-time assertion that sizeof(key_t) > SHA_DIGEST_LENGTH */
+  return digest.key;
+}
+
+
+static void
+initialize_new_semaphore_values(int sem_id, long permissions)
+{
+  unsigned short init_vals[SI_NUM_SEMAPHORES];
+
+  init_vals[SI_SEM_TICKETS] = init_vals[SI_SEM_CONFIGURED_TICKETS] = 0;
+  init_vals[SI_SEM_REGISTERED_WORKERS] = 0;
+  init_vals[SI_SEM_LOCK] = 1;
+
+  if (semctl(sem_id, 0, SETALL, init_vals) == -1) {
+    raise_semian_syscall_error("semctl()", errno);
+  }
+#ifdef DEBUG
+  print_sem_vals(sem_id);
+#endif
+}
+
+static int
+wait_for_new_semaphore_set(key_t key, long permissions)
+{
+  int i;
+  int sem_id = -1;
+  union semun sem_opts;
+  struct semid_ds sem_ds;
+
+  sem_opts.buf = &sem_ds;
+  sem_id = semget(key, 1, permissions);
+
+  if (sem_id == -1){
+      raise_semian_syscall_error("semget()", errno);
+  }
+
+  for (i = 0; i < INTERNAL_TIMEOUT; i++) {
+
+#ifdef DEBUG
+    printf("Waiting for another process to initialize semaphore values, checked: %d times\n", i);
+#endif
+
+    if (semctl(sem_id, 0, IPC_STAT, sem_opts) == -1) {
+      raise_semian_syscall_error("semget()", errno);
+    }
+
+    // If a semop has been performed by someone else, the values must be initialized
+    if (sem_ds.sem_otime != 0) {
+      break;
+    }
+    sleep(1);
```

Comment:
I think you should change this to something more like 10ms. 1 full second added to boot on deploys in a new namespace seems like a bit too much. Most of the time I bet the first worker has been done for a couple of milliseconds before anyone reaches this, but that sucks to prove—so let's just use `usleep(3)`.